### ISSUE TITLE
[a11y] On iOS, dialog titles are announced twice

### ISSUE THREAD

**Issue by jkurtw** [flutter/flutter#44001]
When opening a dialog, on iOS the accessibility focus goes to the first element in the dialog which is the title (incorrect), while on Android the focus goes to the first interactive element, the "Never mind" button, which is correct.

b/143768379

**Comment by darrenaustin** [comment]
Can we have a small example app that shows this? Are they using an `AlertDialog` or something that they built themselves?

**Comment by robertoscaramuzzi** [comment]
We're using `AlertDialog`. I'll try to provide a small repro.

**Comment by Hixie** [comment]
@darrenaustin I've added a tentative target milestone on the assumption that @robertoscaramuzzi can get us a repro case, feel free to adjust as necessary.

**Comment by robertoscaramuzzi** [comment]
So while trying to repro it, I found out that this only happens when the content is wrapped in a `ScrollView`:
```
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Announce Demo',
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  MyHomePage({Key key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('App title'),
      ),
      body: Center(
        child: FlatButton(
          child: Text('Open Dialog'),
          onPressed: () async => await showDialog(
              context: context,
              builder: _buildDialog,
              barrierDismissible: false),
        ),
      ),
    );
  }

  AlertDialog _buildDialog(BuildContext context) => AlertDialog(
        title: Text('Dialog title'),
        content: SingleChildScrollView(
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: [
              Text('Dialog body'),
              FlatButton(
                child: Text('Close'),
                onPressed: () => Navigator.pop(context),
              )
            ],
          ),
        ),
      );
}
```

**Comment by darrenaustin** [comment]
Thanks, @robertoscaramuzzi. I am able to repro the problem with your example. Not sure what is going on here, as the semantics tree only has a single node with the "Dialog title' text. I will see what is going on in the engine and see if I can sort it out.

**Comment by darrenaustin** [comment]
In the engine's iOS a11y bridge, we announce a screen change when the dialog is brought up by posting a `UIAccessibilityScreenChangedNotification` with the dialog's title. Then Voice Over moves the a11y focus to the title, which it then read. Hence the duplication. 

The problem is that we detect screen changes by tracking route changes. The name of the route is what we use to announce the screen change. A lot of the time this is just the title of the dialog or app bar. We could just announce the screen change with no parameter, but then in cases where there is an actual semantic label associated with the route that isn't a title (for example, a popup menu uses "Popup Menu", as it has no title), nothing would be announced.

Not sure how to fix this, as the place that is announcing the screen change doesn't have any way of telling if the route name is meant to be a title or not.

As this issue has been lowered in priority I will move onto more critical a11y issues and come back to this afterwards.

**Comment by chunhtai** [comment]
@darrenaustin Is this issue just waiting for the pr to be merged?

**Comment by darrenaustin** [comment]
> @darrenaustin Is this issue just waiting for the pr to be merged?

Sorry for the delay on responding. It's been a while since I looked a this, and I'm not sure my PR is the right solution. It would solve some common cases, but I wasn't sure if it was a good general solution. I was still looking for feedback at the time that I moved on to other tickets.

**Comment by TahaTesser** [comment]
I can reproduce it 
<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel dev, 1.20.0-7.1.pre, on Mac OS X 10.15.6 19G73, locale
    en-GB)
    • Flutter version 1.20.0-7.1.pre at /Users/taha/Code/flutter_dev
    • Framework revision 7736f3bc90 (6 days ago), 2020-07-10 16:33:05 -0700
    • Engine revision d48085141c
    • Dart version 2.9.0 (build 2.9.0-21.2.beta)

 
[✓] Android toolchain - develop for Android devices (Android SDK version
    30.0.0)
    • Android SDK at /Users/taha/Code/sdk
    • Platform android-30, build-tools 30.0.0
    • ANDROID_HOME = /Users/taha/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google
      Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 47.1.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.47.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.12.2

[✓] Connected device (5 available)
    • SM M305F (mobile)      • 32003c30dc19668f          • android-arm64  •
      Android 10 (API 29)
    • Taha’s iPhone (mobile) • 00008020-001059882212002E • ios            •
      iOS 13.6
    • macOS (desktop)        • macos                     • darwin-x64     •
      Mac OS X 10.15.6 19G73
    • Web Server (web)       • web-server                • web-javascript •
      Flutter Tools
    • Chrome (web)           • chrome                    • web-javascript •
      Google Chrome 83.0.4103.116

• No issues found!
```

</details>

**Comment by mehmetf** [comment]
Customer:money started hitting this claiming that the workaround (making dialogs dismissable) is no longer working. I asked them if they can pinpoint exactly when the workaround stopped working. See internal bug: b/143768379

**Comment by chunhtai** [comment]
The workaround is conflicting with https://github.com/flutter/flutter/issues/46625

**Comment by gaaclarke** [comment]
This problem was working-as-intended as far as the engine is concerned so https://github.com/flutter/engine/pull/19826 is going to be reverted.  Saying the route name and the first item is the correct behavior, in this case they happen to be the same thing.

See https://github.com/flutter/flutter/issues/67345#issuecomment-705851312 for more information.

The correct change should be on the framework side if we don't want them to both be announced.  I tried this one iOS and it works, I'll need to verify on Android:
```diff
diff --git a/packages/flutter/lib/src/material/dialog.dart b/packages/flutter/lib/src/material/dialog.dart
index 17e50a5ad3..5482a7f8b9 100644
--- a/packages/flutter/lib/src/material/dialog.dart
+++ b/packages/flutter/lib/src/material/dialog.dart
@@ -490,7 +490,7 @@ class AlertDialog extends StatelessWidget {
           style: titleTextStyle ?? dialogTheme.titleTextStyle ?? theme.textTheme.headline6,
           child: Semantics(
             child: title,
-            namesRoute: label == null,
+            namesRoute: false,//label == null,
             container: true,
           ),
         ),
```

**Comment by gaaclarke** [comment]
I played around with disabling the `namesRoute` parameter.  Since iOS always selects the first item it is clearly an improvement.  On android when opening the dialog the second time, the last item that was selected is selected which is the close button.  So, talkbalk announces that there is an alert and announces that the close button is selected but the route name isn't announced.  I suspect either behavior is fine, neither would be disorienting.  I would check with someone who actually navigates via TalkBack / VoiceOver.

**Comment by gaaclarke** [comment]
Working as intended, please see https://github.com/flutter/flutter/pull/67700#issuecomment-712438056 for further justification.
